kubernetes-client/src/com/goyeau/kubernetes/client/api/PodsApi.scala

```diff
@@ -71,50 +96,190 @@ private[client] class NamespacedPodsApi[F[_]](
+    val destination = Resource.make(F.delay(new FileOutputStream(destinationFile.toFile)))(s => F.delay(s.close()))
+
+    destination.use { destFile =>
+      wsClient.connectHighLevel(request).use { connection =>
+        connection.receiveStream
+          .through(processWebSocketData)
+          .compile
+          .fold((List.empty[StdErr], none[ErrorOrStatus])) { case ((accEvents, accStatus), data) =>
+            val (errors, status) = data match {
+              case Left(StdOut(data)) =>
+                destFile.write(data)
+                List.empty[StdErr] -> none
+              case Left(e: StdErr) =>
+                List(e) -> none
+              case Right(statusOrError) =>
+                List.empty[StdErr] -> accStatus.orElse(statusOrError.some)
+            }
+
+            (accEvents ++ errors, status)
+          }
+      }
+    }
+  }
+
+  def uploadFile(
+      podName: String,
+      sourceFile: Path,
+      destinationFile: Path,
+      container: Option[String] = None
+  ): F[(List[StdErr], Option[ErrorOrStatus])] = {
+    val destinationDir = Option(destinationFile.getParent)
+
+    val mkDirResult = destinationDir match {
+      case Some(dir) =>
+        val mkDirRequest = execRequest(
+          podName,
+          Seq("sh", "-c", s"mkdir -p $dir"),
+          container
+        )
+        wsClient
+          .connectHighLevel(mkDirRequest)
+          .map(conn => F.delay(conn.receiveStream.through(processWebSocketData)))
+          .use(_.flatMap(foldErrorStream))
+      case None =>
+        (List.empty -> None).pure
+    }
+
+    val uploadRequest = execRequest(
+      podName,
+      Seq("sh", "-c", s"cat - >$destinationFile"),
+      container,
+      stdin = true
+    )
+
+    val uploadFileResult = wsClient.connectHighLevel(uploadRequest).use { connection =>
+      val source = Files[F].readAll(sourceFile, 4096)
+      val sendData = source
+        .mapChunks(chunk => Chunk(WSFrame.Binary(ByteVector(chunk.toChain.prepend(StdInId).toVector))))
+        .through(connection.sendPipe)
+      val retryAttempts = 5
+      val sendWithRetry = Stream
+        .retry(sendData.compile.drain, delay = 500.millis, nextDelay = _ * 2, maxAttempts = retryAttempts)
+        .onError { case e =>
+          Stream.eval(Logger[F].error(e)(s"Failed send file data after $retryAttempts attempts"))
         }
+
+      val result = for {
+        signal <- Stream.eval(SignallingRef[F, Boolean](false))
+        dataStream = sendWithRetry *> Stream.eval(signal.set(true))
+
+        output = connection.receiveStream
+          .through(
+            processWebSocketData
+          )
+          .interruptWhen(signal)
+          .concurrently(dataStream)
+
+        errors = foldErrorStream(
+          output
+        ).map { case (errors, _) => errors }
+      } yield errors
+
+      result.compile.lastOrError.flatten
+    }
+
+    for {
+      result <- mkDirResult
```

Comment:
upload is currently done via two commands, I can't figure out how to make it working using single command with bash '&&'. It simply didn't work